File: app/src/main/java/org/antrack/app/ui/WizardActivity.java
package org.antrack.app.ui;

import android.Manifest;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;

import org.antrack.app.C;
import org.antrack.app.Init;
import org.antrack.app.Pw;
import org.antrack.app.Settings;
import org.antrack.app.libs.Admin;
import org.antrack.app.libs.Battery;
import org.antrack.app.libs.Shell;
import org.antrack.app.libs.Utils;

import java.util.List;

import app.R;

public class WizardActivity extends AppCompatActivity {
    boolean pluginChosen = false;

    private Admin aTools;
    private Pw pw;

    private Button closeButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Dexter.withActivity(this)
                .withPermissions(
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.ACCESS_WIFI_STATE,
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        Manifest.permission.CAMERA,
                        Manifest.permission.RECORD_AUDIO,
                        Manifest.permission.SEND_SMS,
                        Manifest.permission.CALL_PHONE,
                        Manifest.permission.READ_PHONE_STATE,
                        Manifest.permission.PROCESS_OUTGOING_CALLS,
                        Manifest.permission.READ_CONTACTS,
                        Manifest.permission.READ_SMS
                ).withListener(new MultiplePermissionsListener() {
            @Override public void onPermissionsChecked(MultiplePermissionsReport report) {
                if (report.areAllPermissionsGranted()) {
                    // FIXME API
                    Battery.requestIgnoreBatteryOptimisation(WizardActivity.this);
                    main();
                }
            }
            @Override public void onPermissionRationaleShouldBeShown(List<PermissionRequest> permissions, PermissionToken token) {/* ... */}
        }).check();
    }

    private void main() {
        final Context context = this;
        setContentView(R.layout.activity_wizard);

        Init.all(this);

        final Button button_admin = (Button) findViewById(R.id.button_admin);
        button_admin.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                aTools = new Admin(WizardActivity.this);
                aTools.showDialog(WizardActivity.this);
            }
        });


        final Button rootButton = (Button) findViewById(R.id.button_root);
        if (Shell.checkSu()) {
            rootButton.setOnClickListener(new View.OnClickListener() {
                public void onClick(View v) {
                    if (Shell.checkSuRun()) {
                        Settings.put(C.S_USE_ROOT, C.TRUE);
                        Utils.showToast(context, getResources().getString(R.string.root_rights_granted));
                    } else {
                        Utils.showToast(context, "No root rights");
                    }
                }
            });
        } else {
            rootButton.setEnabled(false);
        }

        final Button dropboxButton = (Button) findViewById(R.id.button_dropbox);
        dropboxButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                Settings.put(C.S_PLUGIN, "dropbox");
                try {
                    pluginChosen = true;
                    pw = Pw.getInstance();
                    pw.auth(WizardActivity.this);
                } catch (InterruptedException e) {
                    Utils.showToast(WizardActivity.this, "No internet, try later");
                }
            }
        });

        closeButton = (Button) findViewById(R.id.button_close);
        closeButton.setEnabled(false);
        closeButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                exit();
            }
        });
    }

    private void exit() {
        Settings.wizardComplete(this);
        // Pw is singleton and will be used by service and activity
        pw.connect();
        finish();
    }

    @Override
    public void onBackPressed() {
        if (Settings.get(C.S_TOKEN) == null) {
            // FIXME translate
            Utils.showToast(WizardActivity.this, "Authentication required");
        } else {
            Settings.wizardComplete(this);
            finish();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        // Handle Dropbox plugin auth
        if (pluginChosen) {
            Pw pw = Pw.getInstance();
            String token = pw.resume();
            if (token != null) {
                Settings.put(C.S_TOKEN, token);
                closeButton.setEnabled(true);
            } else {
                // FIXME здесь надо как-то обрабатывать ситуацию если нет токена
            }
        }
    }
}

File: app/src/main/java/org/antrack/app/ui/WizardActivity.kt
@file:Suppress("OVERRIDE_DEPRECATION")

package org.antrack.app.ui

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.Button
import app.R
import org.antrack.app.App
import org.antrack.app.Settings
import org.antrack.app.WIZARD_COMPLETE_FILE
import org.antrack.app.cloud.Cloud
import org.antrack.app.functions.requestIgnoreBatteryOptimisation
import org.antrack.app.functions.toast
import org.antrack.app.functions.touch
import org.antrack.app.libs.Admin
import org.antrack.app.libs.Shell
import java.io.File

class WizardActivity : PermissionsActivity() {
    private val admin = Admin()
    private val appStatus = AppStatus(this)
    private var pluginChosen = false

    private val filesAccessButton: Button
        get() = findViewById(R.id.button_files_access)

    private val backgroundButton: Button
        get() = findViewById(R.id.button_background_work)

    private val adminButton: Button
        get() = findViewById(R.id.button_admin)

    private val rootButton: Button
        get() = findViewById(R.id.button_root)

    private val dropboxButton: Button
        get() = findViewById(R.id.button_dropbox)

    private val closeButton: Button
        get() = findViewById(R.id.button_close)

    companion object {
        private val wizardCompleteFile = App.dataDir + WIZARD_COMPLETE_FILE

        fun launch(activity: Activity, code: Int) {
            val intent = Intent(activity, WizardActivity::class.java)
            activity.startActivityForResult(intent, code)
        }

        fun needToLaunch(): Boolean {
            return !File(wizardCompleteFile).exists()
        }

        fun wizardComplete() {
            File(wizardCompleteFile).touch()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        checkPermissions()
    }

    private fun checkPermissions() {
        requestPermissions(
            arrayOf(
                Manifest.permission.ACCESS_WIFI_STATE,
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.CAMERA,
                Manifest.permission.RECORD_AUDIO,
                Manifest.permission.SEND_SMS,
                Manifest.permission.CALL_PHONE,
                Manifest.permission.READ_PHONE_STATE,
                Manifest.permission.PROCESS_OUTGOING_CALLS,
                Manifest.permission.READ_CONTACTS,
                Manifest.permission.READ_SMS,
                Manifest.permission.POST_NOTIFICATIONS,
            )
        ) { granted ->
            if (granted) {
                main()
            } else {
                toast(R.string.wizard_nopermissions)
                checkPermissions()
            }
        }
    }

    private fun main() {
        setContentView(R.layout.activity_wizard)
        setupFilesAccessButton()
        setupBackgroundWorkButton()
        setupAdminButton()
        setupRootButton()
        setupDropboxButton()
        setupCloseButton()
    }

    private fun setupFilesAccessButton() {
        if (Build.VERSION.SDK_INT < 30) {
            filesAccessButton.visibility = View.GONE
            return
        }

        filesAccessButton.setOnClickListener {
            if (!appStatus.haveAccessToAllFiles) {
                showAllFilesAccessSettingsScreen()
            } else {
                toast(R.string.already_allowed)
            }
        }
    }

    private fun setupBackgroundWorkButton() {
        backgroundButton.setOnClickListener {
            if (!appStatus.isIgnoringBatteryOptimizations) {
                requestIgnoreBatteryOptimisation()
            } else {
                toast(R.string.already_allowed)
            }
        }
    }

    private fun setupAdminButton() {
        adminButton.setOnClickListener {
            if (!admin.isActive) {
                admin.showDialog(this)
            } else {
                toast(R.string.already_allowed)
            }
        }
    }

    private fun setupRootButton() {
        if (!Shell.checkSu()) {
            rootButton.isEnabled = false
            return
        }

        rootButton.setOnClickListener {
            if (Shell.checkSuRun()) {
                toast(R.string.root_rights_granted)
            } else {
                toast(R.string.no_root_right)
            }
        }
    }

    private fun setupDropboxButton() {
        dropboxButton.setOnClickListener { v ->
            Settings.plugin = Cloud.DROPBOX

            try {
                pluginChosen = true
                Cloud.auth(this@WizardActivity, Settings.plugin)
            } catch (e: InterruptedException) {
                toast(R.string.no_internet)
            }
        }
    }

    private fun setupCloseButton() {
        closeButton.isEnabled = Settings.token.isNotEmpty()
        closeButton.setOnClickListener { exit() }
    }

    override fun onBackPressed() {
        if (Settings.token.isEmpty()) {
            toast(R.string.authentication_required)
        } else {
            exit()
        }
    }

    private fun exit() {
        wizardComplete()
        setResult(RESULT_OK, intent)
        finish()
    }

    override fun onResume() {
        super.onResume()

        if (pluginChosen) {
            processToken(Cloud.resume())
        }
    }

    private fun processToken(token: String?) {
        if (token != null) {
            Settings.token = token
            closeButton.isEnabled = true
        } else {
            toast(R.string.authentication_failed)
        }
    }
}
